Bye bye winthreads; relating to #2385

Lord-Kamina opened the pull request:
So, here's a big one...

I removed essentially all instances of hardcoded win32 threads; adding the necessary patches where required.

Additionally, I updated some of the packages as well, most notably boost which is now at 1.71.0 and building properly against pthreads. 🎉

Lord-Kamina commented:
I just noticed I botched commit messages on ilmbase and openEXR; they're versions 2.3.0, not 2.0.3.

tonytheodore commented:
> they're versions 2.3.0, not 2.0.3

You can always reword and force-push, PRs are temporary branches in this project.

A few other comments:
  - Does the boost update work with all its downstream packages? The primary reason it isn't updated is that it has broken those in the past, and they are hard to update themselves
  - please use `git format-patch` style instead of diff-style patches
  - please check your editor settings and/or run `make cleanup-style` to normalise tabs/spaces

Lord-Kamina commented:
> > they're versions 2.3.0, not 2.0.3
> 
> You can always reword and force-push, PRs are temporary branches in this project.

I was going to do that yesterday but was tired and forgot, actually.

> A few other comments:
> 
> * Does the boost update work with all its downstream packages? The primary reason it isn't updated is that it has broken those in the past, and they are hard to update themselves

Haven't tested all of them, probably. Any specific ones I should test?

> * please use `git format-patch` style instead of diff-style patches

I make my patches using `git diff --no-index` because I had no idea git format-patch existed. I assumed it was the same with some comments written on top. õ_õ

> * please check your editor settings and/or run `make cleanup-style` to normalise tabs/spaces

Will do. Should I use tabs or spaces? Or it doesn't matter as long as each file is consistent with itself?

tonytheodore commented:
> Any specific ones I should test?

```
$ make show-downstream-deps-boost
assimp cgal cpp-netlib hyperscan json_spirit libtorrent-rasterbar luabind openscenegraph osgearth pcl pe-util wt xmlwrapp
```

`pcl` and `wt` are the most problematic.

> ...git format-patch existed. I assumed it was the same with some comments written on top. õ_õ

It is very similar - the comments give future maintainers context around what can removed from the patch (especially if those have been submitted upstream). In some cases it provides something close to a workflow where mxe patches are just a series of commits that can be rebased.

> Should I use tabs or spaces?

Always spaces in `src/*.mk` - there are cases in other Makefile rules that `make cleanup-style` may not correct. (my personal preference is always spaces and strip trailing, but preserve actual tabs).

> Or it doesn't matter as long as each file is consistent with itself?

It doesn't matter per-se, review the diffs the PR is generating and avoid generating whitespace changes (unless you are certain).

Lord-Kamina commented:
> > Any specific ones I should test?
> 
> ```
> $ make show-downstream-deps-boost
> assimp cgal cpp-netlib hyperscan json_spirit libtorrent-rasterbar luabind openscenegraph osgearth pcl pe-util wt xmlwrapp
> ```
> 
> `pcl` and `wt` are the most problematic.
> 
> > ...git format-patch existed. I assumed it was the same with some comments written on top. õ_õ
> 
> It is very similar - the comments give future maintainers context around what can removed from the patch (especially if those have been submitted upstream). In some cases it provides something close to a workflow where mxe patches are just a series of commits that can be rebased.
> 
> > Should I use tabs or spaces?
> 
> Always spaces in `src/*.mk` - there are cases in other Makefile rules that `make cleanup-style` may not correct. (my personal preference is always spaces and strip trailing, but preserve actual tabs).
> 
> > Or it doesn't matter as long as each file is consistent with itself?
> 
> It doesn't matter per-se, review the diffs the PR is generating and avoid generating whitespace changes (unless you are certain).

I really need to get more familiar with MXE targets. I ended up running a regex search for  boost among deps and left it running. By default I build most things as shared though so I'll need to try a static build also later. Anyway, if there are build failures you can expect me to either look into them and/or update the problematic packages also.

I've always preferred tabs but as I said my editor can go from one to the other easily so I'll fix that later also.

Lord-Kamina commented:
@tonytheodore I was working further on this over the weekend. I had to make some more changes to these and a few other packages but so far most things are building fine, including pcl and wt.

I'm currently facing some issues with cgal and libtorrent-rasterbar possibly among a few others. Will rebase and push again once I've got them sorted.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

Lord-Kamina commented:
P.S. Jesus. The format patch tool makes work so much more straightforward and easier. I can't believe I was working the badwrong™️ way until now.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

tonytheodore commented:
Using the format-patch style enables a workflow like:
```
# initialise repo based on tarball + patches in $mxe-dir/gits/foo-1.2.3
make init-git-foo
make import-patch-foo

# make changes in gits/foo-1.2.3 or find fix an apply with
curl -L 'https://github.com/foo/foo/commit/1a2b3c4d.patch' | git am

# check changes with
make foo foo_SOURCE_TREE=gits/foo-1.2.3

# commit changes and export patch
make export-patch-foo
make foo

# send PR to mxe
```

Once you get used to it, git handles lots of things nicely.

Lord-Kamina commented:
> Once you get used to it, git handles lots of things nicely.

Yeah that's what I've been doing now that I realized how it works.

Lord-Kamina commented:
This should be good to go now. I've tested, fixed and updated pretty much every thing that depends on boost. I've also updated cmake and even made some fixes (OpenEXR and IlmBase) I've committed PRs upstream for.

It's conflicting with a couple of your updates though, I'm not sure whose commits fixing those issues are better.

For openthreads and openscenegraph, I changed the TRY_RUN to TRY_COMPILE... for openal, I remove and added patches as appropriate, but I also fixed one (mmdevapi is actually an old name for wasapi; I'm not sure if you caught that)

Lord-Kamina commented:
I know this PR is humongous but think you could take a look at merging it at some point?
I checked the conflicts and tries as much as possible to merge the fixes in master with the ones I made.

Lord-Kamina commented:
@tonytheodore think somebody could take a serious look at this PR? even if you don't end-up merging all of it (some parts already might have become obsolete but some solve problems outstanding on master or a few that were solved already but in different ways) but a lot of work went into making everything play nice with pthreads; most of the updates after that are me having noticed certain broken downstream packages and fixing them. 

And, there are many plain-old package updates in here also.
If any build-failures remain I'm committed to fix them, most of the things here I've been using in practice to produce working 32 and 64-bit installers for perfornous.

tonytheodore commented:
Thanks for the updates @Lord-Kamina, I do appreciate the effort that you've put into it and will have a look soon - [it's been a long time coming](https://github.com/mxe/mxe/issues/2258).

Large PRs are hard to merge upfront, even when most the changes could be merged individually without much review. Over time, the only (sane) option is to [use `git diff master` to find the current delta](https://github.com/mxe/mxe/pull/2420#issuecomment-646688621) - I've tried various cherry-picking and rebase strategies in the past and end up in all sorts of trouble.

A couple of pthreads-related cleanups I'll add from a toolchain perspective:
  - pthreads package can be removed ([I only kept it for future testing](https://github.com/mxe/mxe/pull/958#issue-48824996))
  - add independent post-gcc|cmake test packages
  - remove `make show-direct-downstream-deps-pthreads` deps
  - use [`[POSIX | WIN32]_THREADS` variable](https://github.com/mxe/mxe/commit/b8850b2bd0dcedf15c1982b320e4b90139e86492) instead of `MXE_GCC_THREADS` tests

Lord-Kamina commented:
Keep in mind I've been actively using this branch (or at least an offshoot of it), so I can work to solve current conflicts with master.

tonytheodore commented:
There's probably four phases I would approach this in:
  - pthreads cleanup
  - win32threads removal and related updates
  - unrelated updates
  - boost update

It's probably not worth your effort to resolve conflicts since there's also whitespace and style changes, and I'll most likely leave the boost update as a new open PR.

Lord-Kamina commented:
> It's probably not worth your effort to resolve conflicts since there's also whitespace and style changes, and I'll most likely leave the boost update as a new open PR.

Too late.

Also, keep the following in mind: The reason is this PR is so big is that I began changing stuff and seeing downstream packages broken; so I began fixing those and that eventually turned into updating those many times.

The boost update will break many things if you isolate it.

Lord-Kamina commented:
The one thing you'll most likely want to pick-out (unless you changed your mind since I last asked) is that this PR updates the core gcc to gcc8.

tonytheodore commented on src/boost.mk:
```diff
@@ -60,29 +65,48 @@ define $(PKG)_BUILD
+    newlib=`echo \`basename $${lib}\` | $(AWK) '{gsub(/-mt-x[0-9]{2}/,"-mt"); gsub(/_pthread/,"")}1'`; \
+    echo ln -sf "$${lib}" "$(PREFIX)/$(TARGET)/lib/$${newlib}"; \
+    ln -sf "$${lib}" "$(PREFIX)/$(TARGET)/lib/$${newlib}"; \
+    done
```
Don't rename the libs, it's failing here (x86_64-apple-darwin19.5.0):
```
for lib in `ls "/Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib"/libboost_*.a | tr "\n" " "`; do newlib=`echo \`basename ${lib}\` | /opt/local/bin/gawk '{gsub(/-mt-x[0-9]{2}/,"-mt"); gsub(/_pthread/,"")}1'`; echo ln -sf "${lib}" "/Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/${newlib}"; ln -sf "${lib}" "/Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/${newlib}"; done
ln -sf /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_atomic-mt-s-x32.a /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_atomic-mt-s-x32.a
ln -sf /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_chrono-mt-s-x32.a /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_chrono-mt-s-x32.a
ln -sf /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_container-mt-s-x32.a /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_container-mt-s-x32.a
ln -sf /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_context-mt-s-x32.a /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_context-mt-s-x32.a
ln -sf /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_contract-mt-s-x32.a /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_contract-mt-s-x32.a
ln -sf /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_coroutine-mt-s-x32.a /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_coroutine-mt-s-x32.a
ln -sf /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_date_time-mt-s-x32.a /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_date_time-mt-s-x32.a
ln -sf /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_exception-mt-s-x32.a /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_exception-mt-s-x32.a
ln -sf /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_filesystem-mt-s-x32.a /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_filesystem-mt-s-x32.a
ln -sf /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_graph-mt-s-x32.a /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_graph-mt-s-x32.a
ln -sf /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_iostreams-mt-s-x32.a /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_iostreams-mt-s-x32.a
ln -sf /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_locale-mt-s-x32.a /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_locale-mt-s-x32.a
ln -sf /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_log-mt-s-x32.a /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_log-mt-s-x32.a
ln -sf /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_log_setup-mt-s-x32.a /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_log_setup-mt-s-x32.a
ln -sf /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_math_c99-mt-s-x32.a /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_math_c99-mt-s-x32.a
ln -sf /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_math_c99f-mt-s-x32.a /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_math_c99f-mt-s-x32.a
ln -sf /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_math_c99l-mt-s-x32.a /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_math_c99l-mt-s-x32.a
ln -sf /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_math_tr1-mt-s-x32.a /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_math_tr1-mt-s-x32.a
ln -sf /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_math_tr1f-mt-s-x32.a /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_math_tr1f-mt-s-x32.a
ln -sf /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_math_tr1l-mt-s-x32.a /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_math_tr1l-mt-s-x32.a
ln -sf /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_prg_exec_monitor-mt-s-x32.a /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_prg_exec_monitor-mt-s-x32.a
ln -sf /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_program_options-mt-s-x32.a /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_program_options-mt-s-x32.a
ln -sf /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_random-mt-s-x32.a /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_random-mt-s-x32.a
ln -sf /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_regex-mt-s-x32.a /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_regex-mt-s-x32.a
ln -sf /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_serialization-mt-s-x32.a /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_serialization-mt-s-x32.a
ln -sf /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_stacktrace_basic-mt-s-x32.a /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_stacktrace_basic-mt-s-x32.a
ln -sf /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_stacktrace_noop-mt-s-x32.a /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_stacktrace_noop-mt-s-x32.a
ln -sf /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_system-mt-s-x32.a /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_system-mt-s-x32.a
ln -sf /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_test_exec_monitor-mt-s-x32.a /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_test_exec_monitor-mt-s-x32.a
ln -sf /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_thread_pthread-mt-s-x32.a /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_thread-mt-s-x32.a
ln -sf /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_timer-mt-s-x32.a /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_timer-mt-s-x32.a
ln -sf /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_type_erasure-mt-s-x32.a /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_type_erasure-mt-s-x32.a
ln -sf /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_unit_test_framework-mt-s-x32.a /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_unit_test_framework-mt-s-x32.a
ln -sf /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_wave-mt-s-x32.a /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_wave-mt-s-x32.a
ln -sf /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_wserialization-mt-s-x32.a /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib/libboost_wserialization-mt-s-x32.a
# setup cmake toolchain
printf "set(Boost_THREADAPI "pthread")\\n set (Boost_USE_STATIC_LIBS ON)\\n set (Boost_USE_MULTITHREADED ON)\\n set (Boost_NO_BOOST_CMAKE OFF)\\n set (Boost_USE_STATIC_RUNTIME ON)\\n" > '/Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/share/cmake/mxe-conf.d/boost.cmake'
'i686-w64-mingw32.static-g++' -W -Wall -Werror -ansi -U__STRICT_ANSI__ -pedantic '/Users/tonyt/dev/mxe/src/boost-test.cpp' -o '/Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/bin/test-boost.exe' -std='c++11' -lboost_serialization-mt -lboost_thread-mt -lboost_system-mt -lboost_chrono-mt -lboost_context-mt -L'/Users/tonyt/dev/mxe/usr/i686-w64-mingw32.static/lib'
/Users/tonyt/dev/mxe/usr/bin/i686-w64-mingw32.static-ld: cannot find -lboost_serialization-mt
/Users/tonyt/dev/mxe/usr/bin/i686-w64-mingw32.static-ld: cannot find -lboost_thread-mt
/Users/tonyt/dev/mxe/usr/bin/i686-w64-mingw32.static-ld: cannot find -lboost_system-mt
/Users/tonyt/dev/mxe/usr/bin/i686-w64-mingw32.static-ld: cannot find -lboost_chrono-mt
/Users/tonyt/dev/mxe/usr/bin/i686-w64-mingw32.static-ld: cannot find -lboost_context-mt
collect2: error: ld returned 1 exit status
gmake[1]: *** [Makefile:846: build-only-boost_i686-w64-mingw32.static] Error 1
gmake[1]: Leaving directory '/Users/tonyt/dev/mxe'
```

Lord-Kamina replied:
Goddamn it, this is a pain in the ass; I changed it like four times last night and eventually left the one that built fine here. Keep in mind, I'm not renaming, I'm adding symlinks.

But I see the issue, I think; the problem is the "s" static suffix; I think we do in fact need to rename them to remove that.

tonytheodore replied:
I'd be happy to remove the manual test altogether and instruct users to use the provided cmake routines.

Lord-Kamina replied:
That would probably be a more robust solution in the long-term.

Lord-Kamina replied:
If only Boost had decent pkg-config support.

tonytheodore commented on src/boost.mk:
```diff
@@ -4,13 +4,14 @@ PKG             := boost
 $(PKG)_TARGETS  := $(BUILD) $(MXE_TARGETS)
-$(PKG)_DEPS     := cc bzip2 expat zlib
+$(PKG)_DEPS     := cc bzip2 expat zlib xz
+AWK             = $(shell which $(shell gawk --help >/dev/null 2>&1 && echo g)awk)
```
Add such variables near the [top of the main Makefile](https://github.com/mxe/mxe/blob/master/Makefile#L32).

Lord-Kamina replied:
Can't fix it until tomorrow but will do.

tonytheodore commented on src/ffmpeg.mk:
```diff
@@ -37,8 +37,7 @@ define $(PKG)_BUILD
         --disable-debug \
-        --disable-pthreads \
-        --enable-w32threads \
+        --disable-$(if $(findstring posix, $(MXE_GCC_THREADS)),w32threads,pthreads) \
```
Replace `MXE_GCC_THREADS ` test with either `WIN32_THREADS` or `POSIX_THREADS` tests.

Lord-Kamina replied:
Ok

Lord-Kamina replied:
@tonytheodore What about cases where I just want to check threading against no threading? Should I use `$(or)` or will `MXE_GCC_THREADS` be good for those?

tonytheodore replied:
> or will MXE_GCC_THREADS be good for those

`MXE_GCC_THREADS` is compiler-specific, consider the case where it isn't set.

Short-circuit `$(or)` is ideal, though nested `ifs` may be more readable in the future.

tonytheodore replied:
If you're unsure, default to pthreads. 

I'm pretty sure most w32threads are placeholders.

tonytheodore commented on src/ilmbase.mk:
```diff
@@ -4,12 +4,12 @@ PKG             := ilmbase
+$(PKG)_GH_CONF  := AcademySoftwareFoundation/openexr/releases,v
+$(PKG)_FILE     := openexr-v$($(PKG)_VERSION).tar.gz
+$(PKG)_SUBDIR   := openexr-$($(PKG)_VERSION)
+$(PKG)_DEPS     := cc $(BUILD)~cmake
```
$(BUILD)~cmake is added automatically

Lord-Kamina replied:
Ok

tonytheodore commented on src/nsis.mk:
```diff
@@ -4,8 +4,8 @@ PKG             := nsis
-$(PKG)_VERSION  := 3.05
-$(PKG)_CHECKSUM := b6e1b309ab907086c6797618ab2879cb95387ec144dab36656b0b5fb77e97ce9
+$(PKG)_VERSION  := 3.04
+$(PKG)_CHECKSUM := 609536046c50f35cfd909dd7df2ab38f2e835d0da3c1048aa0d48c59c5a4f4f5
```
Move this and the scons to a new PR and base your changes after the parallel build fix.

tonytheodore commented on src/openexr.mk:
```diff
@@ -19,50 +20,16 @@ define $(PKG)_UPDATE
 endef
```
Can this update section be removed?

Lord-Kamina replied:
Yeah, probably

tonytheodore commented on src/openexr.mk:
```diff
@@ -19,50 +20,16 @@ define $(PKG)_UPDATE
 
     '$(TARGET)-g++' \
-        -Wall -Wextra -std=gnu++0x \
+        -Wall -Wextra -std=c++$(if $($(PKG)_CXXSTD_14),14,11) \
```
Where is `$(PKG)_CXXSTD_14` defined?

Lord-Kamina replied:
It's an aborted idea (get gcc versions from gcc.mk, [ which wasn't always actually parsed when these were called ] and compare against version numbers known to provide good standards support for each version) but it was too messy and I forgot to remove it from that file; should be just c++11

tonytheodore commented on src/xmlwrapp.mk:
```diff
@@ -19,10 +17,8 @@ endef
         --host='$(TARGET)' \
-        --disable-shared \
+        $(MXE_CONFIGURE_OPTS) \
         --prefix='$(PREFIX)/$(TARGET)' \
```
remove `--host` and `--prefix` when using `$(MXE_CONFIGURE_OPTS)`

Lord-Kamina replied:
Ok

tonytheodore commented on src/boost.mk:
```diff
@@ -60,29 +65,48 @@ define $(PKG)_BUILD
+    set (Boost_USE_STATIC_LIBS $(if $(BUILD_SHARED),OFF,ON))\\n\
+    set (Boost_USE_MULTITHREADED ON)\\n\
+    set (Boost_NO_BOOST_CMAKE OFF)\\n\
+    set (Boost_USE_STATIC_RUNTIME $(if $(BUILD_SHARED),OFF,ON))\\n" > '$(CMAKE_TOOLCHAIN_DIR)/$(PKG).cmake'
```
Indentation

tonytheodore commented on src/ilmbase.mk:
```diff
@@ -19,23 +19,11 @@ define $(PKG)_UPDATE
+    -DOPENEXR_CXX_STANDARD=11 \
+    -DBUILD_TESTING=OFF \
+    -DILMBASE_INSTALL_PKG_CONFIG=ON \
+    "$(SOURCE_DIR)"/IlmBase
```
Indentation

tonytheodore commented on src/openexr.mk:
```diff
@@ -19,50 +20,16 @@ define $(PKG)_UPDATE
+    -DOPENEXR_CXX_STANDARD=11 \
+    -DOPENEXR_INSTALL_PKG_CONFIG=ON \
+    -DBUILD_TESTING=OFF \
+    "$(SOURCE_DIR)/OpenEXR"
```
Indentation

Lord-Kamina replied:
Should be fixed now.

Lord-Kamina commented:
Only thing left from the earlier comments is picking scons and nsis out; will do it after I finish having lunch

Lord-Kamina commented:
Hopefully I didn't botch anything with that merge.

tonytheodore commented:
> Hopefully I didn't botch anything with that merge.

Apart from now having 247 commits instead of the previous 30 odd? You need to make it fast-forward friendly.

The changes themselves look good and I've kicked off some test builds.

Lord-Kamina commented:
> > Hopefully I didn't botch anything with that merge.
> 
> Apart from now having 247 commits instead of the previous 30 odd? You need to make it fast-forward friendly.

That was sadly a by-product of excising the nsis thing.

Lord-Kamina commented:
Ugh I think I broke it; I made a backup just before the last attempt at a rebase though. Will try to sort this crap tomorrow.

Lord-Kamina commented:
@tonytheodore  should be a lot cleaner now. :)

tonytheodore commented:
Failed packages from last test using gcc8, (ignore librsvg and ocaml - they're failing for unrelated reasons):
```
drwxr-xr-x  5 tonytheodore mxe    4096 Jun 24 06:57 tmp-assimp-i686-w64-mingw32.static
drwxr-xr-x  5 tonytheodore mxe    4096 Jun 24 07:03 tmp-assimp-x86_64-w64-mingw32.static
drwxr-xr-x  5 tonytheodore mxe    4096 Jun 24 17:57 tmp-cgal-i686-w64-mingw32.shared
drwxr-xr-x  6 tonytheodore mxe    4096 Jun 24 15:02 tmp-cgal-i686-w64-mingw32.static
drwxr-xr-x  5 tonytheodore mxe    4096 Jun 24 22:25 tmp-cgal-x86_64-w64-mingw32.shared
drwxr-xr-x  6 tonytheodore mxe    4096 Jun 24 17:57 tmp-cgal-x86_64-w64-mingw32.static
drwxr-xr-x  5 tonytheodore mxe    4096 Jun 24 03:10 tmp-db-i686-w64-mingw32.shared
drwxr-xr-x  5 tonytheodore mxe    4096 Jun 24 08:01 tmp-db-x86_64-w64-mingw32.shared
drwxr-xr-x  5 tonytheodore mxe    4096 Jun 24 18:36 tmp-gtkmm2-i686-w64-mingw32.static
drwxr-xr-x  5 tonytheodore mxe    4096 Jun 24 22:51 tmp-gtkmm2-x86_64-w64-mingw32.static
drwxr-xr-x  5 tonytheodore mxe    4096 Jun 24 22:59 tmp-gtkmm3-i686-w64-mingw32.shared
drwxr-xr-x  5 tonytheodore mxe    4096 Jun 24 18:39 tmp-gtkmm3-i686-w64-mingw32.static
drwxr-xr-x  5 tonytheodore mxe    4096 Jun 25 01:35 tmp-gtkmm3-x86_64-w64-mingw32.shared
drwxr-xr-x  5 tonytheodore mxe    4096 Jun 24 23:00 tmp-gtkmm3-x86_64-w64-mingw32.static
drwxr-xr-x  6 tonytheodore mxe    4096 Jun 24 16:05 tmp-imagemagick-i686-w64-mingw32.static
drwxr-xr-x  6 tonytheodore mxe    4096 Jun 24 18:47 tmp-imagemagick-x86_64-w64-mingw32.static
drwxr-xr-x  5 tonytheodore mxe    4096 Jun 24 21:59 tmp-librsvg-i686-w64-mingw32.shared
drwxr-xr-x  5 tonytheodore mxe    4096 Jun 25 01:19 tmp-librsvg-x86_64-w64-mingw32.shared
drwxr-xr-x  5 tonytheodore mxe    4096 Jun 24 12:49 tmp-ocaml-core-i686-w64-mingw32.static
drwxr-xr-x  5 tonytheodore mxe    4096 Jun 24 15:56 tmp-opencv-i686-w64-mingw32.static
drwxr-xr-x  5 tonytheodore mxe    4096 Jun 24 18:15 tmp-opencv-x86_64-w64-mingw32.static
drwxr-xr-x  5 tonytheodore mxe    4096 Jun 24 16:17 tmp-ossim-i686-w64-mingw32.shared
drwxr-xr-x  5 tonytheodore mxe    4096 Jun 24 20:07 tmp-ossim-x86_64-w64-mingw32.shared
drwxr-xr-x  5 tonytheodore mxe    4096 Jun 25 03:37 tmp-pcl-i686-w64-mingw32.shared
drwxr-xr-x  5 tonytheodore mxe    4096 Jun 25 03:36 tmp-pcl-i686-w64-mingw32.static
drwxr-xr-x  5 tonytheodore mxe    4096 Jun 25 03:56 tmp-pcl-x86_64-w64-mingw32.shared
drwxr-xr-x  5 tonytheodore mxe    4096 Jun 25 03:37 tmp-pcl-x86_64-w64-mingw32.static
drwxr-xr-x  5 tonytheodore mxe    4096 Jun 24 13:03 tmp-xmlwrapp-i686-w64-mingw32.shared
drwxr-xr-x  5 tonytheodore mxe    4096 Jun 24 10:23 tmp-xmlwrapp-i686-w64-mingw32.static
drwxr-xr-x  5 tonytheodore mxe    4096 Jun 24 13:32 tmp-xmlwrapp-x86_64-w64-mingw32.shared
drwxr-xr-x  5 tonytheodore mxe    4096 Jun 24 13:04 tmp-xmlwrapp-x86_64-w64-mingw32.static
```

I've kicked off another build, again with gcc8.

Lord-Kamina commented:
Could you please upload logs somewhere? I'll probably have some time to work on this on the weekend.

Lord-Kamina commented:
I'm currently running builds of all the packages you mentioned; will report results later.

tonytheodore commented on src/assimp.mk:
```diff
@@ -11,16 +11,15 @@ $(PKG)_DEPS     := cc boost minizip
+        -W -Wall -Werror -ansi -pedantic -std=c++11 \
         '$(TEST_FILE)' -o '$(PREFIX)/$(TARGET)/bin/test-assimp.exe' \
-        `'$(TARGET)-pkg-config' assimp minizip --cflags --libs`
+        `'$(TARGET)-pkg-config' assimp minizip --cflags --libs` -lirrxml
```
The lib is named `libIrrXML` and fails on case-sensitive systems for all targets.

Also, the shared builds don't install a *.dll version of IrrXML.

Lord-Kamina replied:
> zstd was recently added to mxe and can be added to DEPS and enabled
> 
> Edit: that applies to minizip

I know, I explicitly disabled it though because the build fails at linking, maybe in the future it will work properly.

Lord-Kamina replied:
Apparently this was a bug upstream. I'm currently investigating more.

Lord-Kamina replied:
So.. the shared version of assimp links IrrXML statically, which is why it obviously doesn't install it (also it's a third-party lib so why would they install it?) but now I'm trying to understand why the static version DOES install it.

tonytheodore commented on src/cgal.mk:
```diff
@@ -28,13 +29,15 @@ define $(PKG)_BUILD
     cd '$(BUILD_DIR).test-image' && '$(TARGET)-cmake' '$(SOURCE_DIR)/examples/CGALimageIO' \
-        -DCMAKE_CXX_STANDARD=98
+        -DCMAKE_CXX_STANDARD=11 \
+        -C'$(PWD)/src/cgal-TryRunResults.cmake'
```
This fixes cgal on all targets.

tonytheodore commented:
pcl has several updates that remove most of the boost classes. I'm testing the latest updates against master and it's probably not worth investigating the failed detection.

tonytheodore commented on src/gtk2.mk:
```diff
@@ -19,7 +19,17 @@ define $(PKG)_UPDATE
+
 define $(PKG)_BUILD
+    $(if $(BUILD_STATIC), $(call $(PKG)_PATCH_STATIC))
+    $(if $(findstring x86_64, $(TARGET)), $(call $(PKG)_PATCH_X64))
```
Variable patches for a single version are [deliberately excluded from the pkg rule lookup](https://github.com/mxe/mxe/blob/4ccfdfc560c63188c1c61a1abbb9e736cc985715/Makefile#L572). There's never a good reason for them:
  - can't be sent upstream
  - hard to maintain
  - don't fit with patch workflow
  - there's always a correct set of defines and macros to achieve the same result

The latter is, of course, a lot of work. It's better to leave features disabled till someone is prepared to do that work.

Lord-Kamina replied:
> * there's always a correct set of defines and macros to achieve the same result
> 
> The latter is, of course, a lot of work. It's better to leave features disabled till someone is prepared to do that work.

Trust me, there isn't. I tried fixing them before I did this.
Several of the current patches in master I moved to the static patch because they were actively breaking the shared builds. By contrast, the x64 build waw always going to fail because of a lot of compatibility code thought out by them to play nice with older versions of tye the other library but we don't really care about that.

Keep in mind that this is gtk2 we're talking about. It's essentially an EoL package already.
The x64 patch was pulled from a commit under the `2.90.7` tag upstream.

Lord-Kamina replied:
P.S. I went into the #gtk official channel at one point asking if they could help me solve some of the errors. Their literal reply? "Why do you still want gtk2 anyway?"

Lord-Kamina commented:
@tonytheodore does ossim build in master? Because that library looks like a steaming hot mess...

tonytheodore commented:
> does ossim build in master?

Apologies, I missed that question. 

Yes, last time I checked (about two weeks ago), all packages (including ossim) were building with gcc5 and gcc9. gcc10 has one failing package, qtifw, and also needs nsis revised with the final upstream fix.

Lord-Kamina commented:
What's the status of the various packages on master? I'd really like to try again to eventually get this merged.

danger89 commented:
This will cause merge conflicts with the soon upcoming PR: https://github.com/danger89/mxe/pull/1 (Updating the whole GTK Stack). And maybe making this PR be superfluous.

Lord-Kamina commented:
> And maybe making this PR be superfluous.
> 

I doubt it, considering this touched upon many things other than GTK.

Lord-Kamina commented:
Conflicts solved again.

danger89 commented:
I agree. This needs to be merged.
